Deprecate logcfg

melanieclarke opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Partially resolves [JP-1866](https://jira.stsci.edu/browse/JP-1866)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #244 
Closes #132 
Closes #185 
Closes #241 

<!-- describe the changes comprising this PR here -->
Deprecate usage of log configuration via `logcfg`.  Users should move to using command line arguments for setting the log level or filename, or configure logging directly in their Python code.

For the transition period, logcfg-style configuration is still available, but it will raise a DeprecationWarning if it is used from the command line, in `call` function arguments, or implicitly via "stpipe-log.cfg" files.

Default configuration for the command line or the call function is still to log INFO and higher messages to stderr.  It is not applied in the call function if handlers have already been added to the root logger, or if `configure_log=False` is passed in the call function keyword arguments.

Changes to downstream packages are not required before merge, but JWST unit and regression tests will fail due to the Deprecation Warning until a PR is merged to avoid using logcfg.  That PR will require the changes here to support new log configuration arguments, so it should be coordinated with this PR.

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] run regression tests with this branch installed (`"git+https://github.com/<fork>/stpipe@<branch>"`)
    - [ ] [`jwst` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml)
    - [ ] [`romancal` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.feature.rst``: new feature
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.doc.rst``: documentation change
- ``changes/<PR#>.removal.rst``: deprecation or removal of public API
- ``changes/<PR#>.misc.rst``: infrastructure or miscellaneous change
</details>

melanieclarke commented:
Regression tests for jwst on deprecate_logcfg branch:
https://github.com/spacetelescope/RegressionTests/actions/runs/16376272705

Regression tests for romancal on main:
https://github.com/spacetelescope/RegressionTests/actions/runs/16376314902

penaguerrero reviewed (approved):
I tested these changes and saw no issues. LGTM

melanieclarke commented on src/stpipe/cmdline.py:
```diff
@@ -61,6 +65,75 @@ def _get_config_and_class(identifier):
+            help="Turn on all logging messages",
+        )
+        parser1.add_argument(
+            "--log_level",
```
Renamed in [6dcd0ae](https://github.com/spacetelescope/stpipe/pull/250/commits/6dcd0ae70b88f77fab69435e076f9e58818a02b8)

I updated the jwst PR to match.

braingram commented on tests/test_logger.py:
Thanks! Excellent coverage.

braingram commented on changes/250.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Provide new command line arguments for configuring logging (log_level, log_format, log_file, log_stream).
```
```suggestion
Provide new command line arguments for configuring logging (log-level, log-format, log-file, log-stream).
```

braingram commented on src/stpipe/cmdline.py:
```diff
@@ -15,6 +16,9 @@
     "debug",
     "logcfg",
     "verbose",
+    "log_level",
+    "log_file",
+    "log_stream",
```
```suggestion
    "log-level",
    "log-file",
    "log-stream",
```

melanieclarke replied:
Thanks, I missed that one.

melanieclarke commented on src/stpipe/step.py:
```diff
@@ -731,16 +732,36 @@ def call(cls, *args, **kwargs):
+            # of always configuring the root logger, we should
+            # check all known loggers here as well.
+            root_logger = logging.getLogger()
+            if not log.is_configured(root_logger):
```
I'd like to discuss this some more -- this behavior is an intended design feature for Python users, not an error.

The idea is that we only apply the default configuration as a fallback, if the user has not set up their own handling via the logging module.  If they have, then we assume they have set it up they way they want it and we should not add anything more to it.  Essentially, our default logging configuration should be a logger of last resort, rather than something you have to know how to explicitly turn off if you need something different in your code.

I think this is important to enable users to interact with pipeline code in a natural, Pythonic way, using the logging module only.

melanieclarke replied:
Well, not quite -- they would see logging messages as configured by the other package.  This would be pretty poor behavior on the part of the other package, exactly the sort of thing we are trying to move away from in stpipe.

The currently implemented behavior is documented for the `call` function -- is that an insufficient explanation, if the user does run into a confusing situation?

braingram replied:
They would only see the log messages if that other handler displays them. I think it's a mistake to rely on behavior of other packages to decide if we display log messages when we can just display them or the user can disable them.

Thanks for pointing out the documentation would you also remove that note?

melanieclarke replied:
Yes, of course if we decide to remove the code, I will also remove the note.

But I still disagree that the user should have to explicitly disable log configuration every time they run the `call` function, rather than setting up their own logging configuration once and expecting that the pipeline code will not override it.  This is not relying on other packages to configure logging, this is relying on the user, in their own code, to set things up as they desire. 

For example, if I am running many pipeline steps in a notebook, and I'd like to direct log messages to a file and not print messages to the screen, I'd like to set that up once, at the top of the notebook.  If we insist on adding a default logging configuration every time, I will have to explicitly turn it off for every pipeline step.

Would you feel differently about this if we were checking only the stpipe loggers for added handlers, as in #255, rather than the root logger?

braingram replied:
> Yes, of course if we decide to remove the code, I will also remove the note.
> 
> But I still disagree that the user should have to explicitly disable log configuration every time they run the `call` function, rather than setting up their own logging configuration once and expecting that the pipeline code will not override it. This is not relying on other packages to configure logging, this is relying on the user, in their own code, to set things up as they desire.
> 
> For example, if I am running many pipeline steps in a notebook, and I'd like to direct log messages to a file and not print messages to the screen, I'd like to set that up once, at the top of the notebook. If we insist on adding a default logging configuration every time, I will have to explicitly turn it off for every pipeline step.

Thanks for the details. I understand what you're saying but I disagree both on what's expected here and how it's implemented.

As far as I know it's always been the case that by default `Step.call` logs to the console. The change here is to add a check that if a logger (either root or one of the named loggers in #255) has any handler (except for a FileHandler writing to /dev/null, or a LogCaptureHandler or a _LiveLoggingNullHandler) `Step.call` will no longer display log messages. I'm not in favor of that change in default behavior.

The "_LiveLoggingNullHandler" in the code was what originally led me down this route. It's a private log name from pytest. If pytest decides to change this name we will suddenly stop seeing log messages in test runs with no indication that logging was disabled.

If there is a use case for a more global logging configuration why don't we keep logcfg?
 
> Would you feel differently about this if we were checking only the stpipe loggers for added handlers, as in #255, rather than the root logger?

Thanks for asking. No the check for handlers and silently disabling the logging isn't specific to the root logger.

braingram replied:
> I do see your point about the pytest logger. It is possible the root logger might have handlers attached to it that the user is not aware of, but it seems pretty unlikely to me that the specific packages we've named as stpipe loggers would, without our knowledge. If we merge #255 and check for handlers on known loggers only, would that be a reasonable path forward?

Thanks. I don't think we should be silently not applying a configuration based on the state of an external logger (even non-root loggers).
 
> If not, my goal is to have at least one way to offer to users to configure their own logging with the logging module without the call function overriding it by default. Do you see a different solution?

This is a rather major change to how stpipe behaves and at the very least a change like this (where logging output is no longer shown if a logger has any handler) requires a deprecation period where things continue to behave as usual (showing logging output) before a change is made.

Since there is disagreement about the change where `call` no longer shows log output of a logger has an attached handler and this PR is doing multiple things (adding `log-file` etc arguments, deprecating log cfg and changing when the default is applied) would you split this into 3 PRs if any of this work is required for the other logging PRs? At the moment I'm not in favor of the change in when the default configuration is applied but am in favor of the `log-file` etc arguments. I'm also not sure deprecating `logcfg` is an improvement as it removes the ability to globally apply a logging configuration (and the alternative requires a brittle dependence on logger state).

melanieclarke replied:
Doesn't the operational environment use command line interface?  In that use case, logging is never applied internal to the call function.  It is applied directly by the script.  The choices internal to the call function should be relevant only to Python users.

zacharyburnett commented:
> A note in case anyone else is inclined to think (like I do) that we should be erring on the side of not adding handlers in the call function: I'd also be happy to remove the exception for the pytest-style root logger handlers, if that seems too arbitrary/obscure. In that case, it might be useful to add a pytest fixture for jwst and romancal tests that adds a stream handler to the known loggers at the beginning of every test session.
> 
> Also, another note about usability for the call function, related to the tickets mentioned above: I'd be happy to add convenience keywords for log configuration to the call step, if that sounds like a useful shortcut for casual Python users who don't need/want to engage with the logging module. I could add the same arguments we use for the command line interface: log_level, log_stream, log_file, verbose.

IMO the logging module is pretty well documented, so I would say that we _should_ be offloading this as much as possible (and not complicating the CLI more)

melanieclarke commented:
Now that #255 has been merged, I pushed up what I believe to be a pretty safe compromise for the default behavior of the `call` function when called by Python users.  It now checks only the stpipe loggers for existing handlers: if any are present, the default log configuration is not added.  I moved the `is_configured` function back to the testing suite since we don't need the pytest handler exception if we're not checking the root logger.

braingram commented:
> Now that #255 has been merged, I pushed up what I believe to be a pretty safe compromise for the default behavior of the `call` function when called by Python users. It now checks only the stpipe loggers for existing handlers: if any are present, the default log configuration is not added. I moved the `is_configured` function back to the testing suite since we don't need the pytest handler exception if we're not checking the root logger.

Thanks for updating this. I still do not think we should skip configuring the logging based on the state of the loggers and instead make it explicit (either disabling logging configuration by passing `configure_log=False` as is possible with this PR or perhaps adding a global stpipe option to control this).

To provide a relevant example. During the [romancal PR adding `get_stpipe_loggers`](https://github.com/spacetelescope/romancal/pull/1898) it was mentioned that capturing logs from [roman_photoz](https://github.com/spacetelescope/roman_photoz) was an expected future change. However roman_photoz configures it's [own logger](https://github.com/spacetelescope/roman_photoz/blob/951775a086b716c266446e60ec28599866c03640/roman_photoz/logger.py#L7). Testing locally if I add `roman_photoz` to `get_stpipe_loggers` I no longer see any log output when `call` is used after `roman_photoz` is imported (where it configures it's logger).

braingram commented:
> Okay, if the call function must log, it must always log! I have removed the check for prior configurations, and users who want to use the logging module to configure the log will need to set configure_log=False.

Thanks! I'll try to take a detailed look today.

> Agreed, this PR no longer addresses those issues, since configuration for debug level or a log file requires a non-default keyword. Would you prefer that I remove the keywords or address those issues here by extending the command line keywords to the call function?

I'm leaning towards removing the closing keywords to not further delay this PR. Let me know if you'd rather go with the other option. Both are good for me.

melanieclarke commented:
Updated regression tests for jwst, with https://github.com/spacetelescope/jwst/pull/9670 to address tests that use logcfg:
https://github.com/spacetelescope/RegressionTests/actions/runs/17861737228

Updated regression tests for romancal on main:
https://github.com/spacetelescope/RegressionTests/actions/runs/17861776422

braingram reviewed (approved):
Thanks for all your work on this. LGTM.

ddavis-stsci reviewed (approved):
I don't see any issues.

melanieclarke commented:
Thanks for reviewing all!  This PR needs to be coordinated with the jwst PR to avoid test failures, so I will hold off on merging here until that one is approved.
